src/common/options/crimson.yaml.in

```diff
@@ -31,6 +31,13 @@ options:
   desc: CPU cores on which alienstore threads will run in cpuset(7) format
   flags:
   - startup
+- name: crimson_seastar_reactor_num
+  type: uint
+  level: advanced
+  default: 0
+  desc: number of CPU cores on which seastar reactor threads will run
+  flags:
+  - startup
```

Comment:
Not sure why `crimson_seastar_reactor_num` is needed.

Looks it is identical to configure `crimson_seastar_cpu_cores` as "0-`crimson_seastar_reactor_num-1`" ?

Which seems not useful when there are multiple OSDs because `--cpuset` always start with 0, causing overlap.

Reply:
I agree that it's practically the same. It's there only to make Crimson deployment friendlier by setting a global (non-optimal) OSD value of `crimson_seastar_reactor_num = n`.
Once `crimson_seastar_cpu_cores` and `crimson_alien_thread_cpu_cores` are set, this value will no longer have any impact.

Another reason is that if `crimson_seastar_cpu_cores` is not set, we fallback to Seastar's default of using all available cores. That means that on nodes with 80 cores we will create 80 seastar threads.

By looking again, `crimson_seastar_thread_num` is probably a better terminology.

Reply:
I think I generally understand the purpose.

> Which seems not useful when there are multiple OSDs because --cpuset always start with 0, causing overlap.

But I'm still wondering how to prevent "--cpuset" overlapping if there are multiple OSDs in the same host, only with `crimson_seastar_thread_num` or `crimson_seastar_reactor_num` configured.

I suppose the OSD still needs to know that it is the nth daemon in the same host to produce the accurate `--cpuset` for seastar. But if it requires yet another Ceph configuration (which still has to be configured separately), probably the current `crimson_seastar_cpu_cores` is simpler?

Or, does each OSD live in an separate container, so that overlapping is impossible?

Reply:
> But I'm still wondering how to prevent "--cpuset" overlapping if there are multiple OSDs in the same host, only with `crimson_seastar_thread_num` or `crimson_seastar_reactor_num` configured.

It's expected to overlap. As mentioned, it's not optimal (at all), but it's just there to let the cluster come up in a simple way. Actual usage will have to set `crimson_seastar_cpu_cores` and `crimson_alien_thread_cpu_cores`.
I'll emphasis that in the docs.

Reply:
@cyx1231st, Please see updated commit

Reply:
The difference is that if all OSDs on a host specify 0-N for crimson_sesatar_cpu_cores they will all be limited to exactly the same set of cpus -- that's basically always undesirable.  With crimson_seastar_reactor_num, we don't set a cpu_mask at all so the scheduler is free to place them whereever.  In that case, they *may* overlap, but at least all of the cpus on the host will be used.  Operators interested in performance will need to specify crimson_seastar_cpu_cores for every OSD individually, but operators who simply want to try to it out can specify crimson_seastar_reactor_num a single time globally and get reasonable if suboptimal performance.  Such operators can then choose to switch per-osd crimson_seastar_cpu_cores if they wish.

Perhaps it would make sense to send a warning to the central log if a crimson-osd starts without crimson_seastar_cpu_cores (and the corresponding bluestore ones for bluestore) set to alert operators that they may want to make changes for better performance?

Reply:
@athanatos yeah, makes sense to me.

Probably `--smp` and `--thread-affinity=0` can be useful for `crimson_seastar_num_threads`.

The same warning as well as how the 3 configs impact alienstore thread pinning should also be good in the configuration description. User may not look at the comments in the source code.

Reply:
>Perhaps it would make sense to send a warning to the central log if a crimson-osd starts without crimson_seastar_cpu_cores (and the corresponding bluestore ones for bluestore) set to alert operators that they may want to make changes for better performance?

@athanatos, Does that qualify as a health check alert?

---

>Probably --smp and --thread-affinity=0 can be useful for crimson_seastar_num_threads.

* @cyx1231st, What do you suggest to change with `--smp`? Currently we only allow 1 (seastar`s default of one thread per CPU).

* `--thread-affinity=0` may be a good idea for  `crimson_seastar_num_threads`.

>The same warning as well as how the 3 configs impact alienstore thread pinning should also be good in the configuration description. User may not look at the comments in the source code.

Ack, will update the desc. Docs will be updated in a followup PR.

Reply:
@Matan-B No, we don't want to force users to deal with it, merely notify them.  Perhaps just printing a local log message at info would be fine?

Reply:
`-c [ --smp ] arg                      number of threads (default: one per CPU)`

Looks it configures the seastar thread number, and `thread-affinity=0` to unpin the CPUs.

Reply:
See reactor.cc:
```
    if (smp_opts.smp) {
        smp::count = smp_opts.smp.get_value();
    } else {
        smp::count = cpu_set.size();
    }
    std::vector<reactor*> reactors(smp::count);
```
All that `--smp` does is replace cpu_set.size() usage explicitly. Setting `--smp` = n or `--cpuset` = 0-n-1 will result in the **same** number of smp::count. The only difference would be which CPUs are used.

Reply:
I agree that it makes sense to allow the user to specify a number of alienstore cpu threads different from the number of seastar threads.  To make things simpler, I'd be fine with requiring the user to specify both.

Reply:
### Option 1 (Optimal)
* CPUs are pinned to the provided sets (seastar and alien cpu sets)
* thread-affinity = 1 (seastar's default).
* smp (smp::count / reactor count): `seastar_cpu_set.size()`
* Seastar reactor threads: `seastar_cpu_set.size()`
* Alienstore threads: `alien_cpu_set.size()` * num_threads_per_core (default = 4).
--- 
### Option 2 (Basic):
* No CPU pinning.
* thread-affinity = 0
* smp (smp::count / reactor count) = `crimson_seastar_num_threads`
* Seastar reactor threads num: `crimson_seastar_num_threads`
* Alienstore threads num: `crimson_alien_op_num_threads`